### DISCUSSION TITLE
Behavior of cluster_pairwise_predictions_at_threshold

### DISCUSSION THREAD

**Discussion by sthamodh** [moj-analytical-services/splink#1868] [Q&A]
Hi,

I get the following ValueError every once in a while I run the same splink deduplication code. When I re-run the code, it works without any errors. Why would this be the case?

ValueError: You've asked to drop table __splink__df_representatives_f6c11324c from your database which is not a table created by Splink.  If you really want to drop this table, you can do so by setting force_non_splink_table=True

**Comment by RobinL** [comment]
Thanks for the report.  I'd be grateful for a few more details:
- What Splink version are you using?
- What database backend (duckdb etc.)
- Any other details you can provide, ideally a reproducible script

**Reply by sthamodh** [reply]
Thanks Robin,

What Splink version are you using? 
3.9.11

What database backend (duckdb etc.)
SparkLinker.
I work on a databrticks 12.2 ML Runtime cluster if that helps


Any other details you can provide, ideally a reproducible script

`## SELECT required columns from the input table and store them as a delta table in the project sandbox
    df = spark_session.sql(f"""Select
            monotonically_increasing_id() as unique_id,
            {column_selection_sql},
            {cmhc_streetnumber_string} as cmhc_streetnumber_string,
            substr(replace(cmhc_postalcode, ' ', ''), 1, 5) as cmhc_postalcode_5
            from
            {project_sandbox}.{input_table}
            where
            {input_query_non_null_condition}
            {add_sql}
            """).repartition(1500).cache()

    num_rows = df.count()

    spark_session.sql("SET spark.databricks.delta.preview.enabled=true")

    print(f"There are {num_rows} records in total.")

    df.write.mode("overwrite").saveAsTable(f"{project_sandbox}.{input_table_with_uid}")

    ## Adding comparison levels as dictionaries to a list  
    ## The resultant list is input as the value for the key, "comparisons" to the linker's settings argument 
    comparison_list = []


    ## 1. Adding all exact match comparison columns by lopping through the items in the exact_match_column_comparison_dict dictionary
    for comparison_column, comparison_settings in exact_match_column_comparison_dict.items():
        
        var = "comparison_" + comparison_column
        tfa = comparison_settings.get("term_frequency_adjustments", False)
        m_prob = comparison_settings.get("m_probability", None)
            
        cur_dict = {
        "output_column_name": comparison_column,
        "comparison_description": comparison_column + " exact match",
        "comparison_levels": [
            cll.null_level(comparison_column),
            cll.exact_match_level(comparison_column,  term_frequency_adjustments = tfa, m_probability = m_prob),
            cll.else_level(),
            ],}
        
        comparison_list.append(dict(cur_dict))
    
    ## 2. Adding all jaro winkler match comparison columns by lopping through the items in the jaro_winkler_column_comparison_dict dictionary
    for comparison_column, comparison_settings in jaro_winkler_column_comparison_dict.items():
        
        var = "comparison_" + comparison_column
        threshold = comparison_settings.get("threshold", False)
            
        cur_dict = {
        "output_column_name": comparison_column,
        "comparison_description": comparison_column + " jaro winkler match",
        "comparison_levels": [
            cll.null_level(comparison_column),
            cll.jaro_winkler_level("cmhc_streetname", distance_threshold = 0.95),
            cll.else_level(),
            ],}

        comparison_list.append(dict(cur_dict))


    
    
    ## Initialize the settings object for splink's spark linker
    settings = {
        "link_type": "dedupe_only",
        "probability_two_random_records_match": model_params.get("probability_two_random_records_match", None),
        "comparisons": comparison_list,
        "blocking_rules_to_generate_predictions": [" and ".join(blocking_rule)],
        "retain_matching_columns": True,
        "retain_intermediate_calculation_columns": True,
        "em_convergence": model_params.get("em_convergence", None)
    }

    ## Initiliaze the spark linker object for finding duplicates
    linker = SparkLinker(df, settings, spark=spark_session)

    ## Estimate u probabilities using splink's estimate_u_using_random_sampling method
    ## https://moj-analytical-services.github.io/splink/linker.html?h=estimate_u#splink.linker.Linker.estimate_u_using_random_sampling
    linker.estimate_u_using_random_sampling(max_pairs = model_params.get("max_pairs_for_estimating_u", None)
                                            ,seed = model_params.get("seed", None)
                                            )

    ## Estimate m probabilities using splink's estimate_parameters_using_expectation_maximisation method
    ## https://moj-analytical-services.github.io/splink/linker.html?h=estimate#splink.linker.Linker.estimate_parameters_using_expectation_maximisation
    training_blocking_rule_list = model_params.get("training_blocking_rule_list", None)
    
    for rule in training_blocking_rule_list:
        linker.estimate_parameters_using_expectation_maximisation(rule)

    ## display the parameters chart
    display(linker.m_u_parameters_chart())
    ## display the weights chart
    display(linker.match_weights_chart())

    ## Store the linker object's prediction results with a threshold probability
    results = linker.predict(threshold_match_probability = model_params.get("threshold_match_probability", None))
    pdf = linker.cluster_pairwise_predictions_at_threshold(results, model_params.get("threshold_match_probability", None)).as_pandas_dataframe()

    ## save linker model as a dict
    model_dict = linker.save_model_to_json("my_settings.json", overwrite=True)

    ## Write pairwise predictions as a table to the project sandbox
    results = results.as_spark_dataframe()
    results.write.mode("overwrite").saveAsTable(f"{project_sandbox}.tmp_test_pairwise_dups")`

**Comment by pallyndr** [comment]
Same error.
Splink 4.0.4
Backend is Spark on Databricks.
Error is sporadic.

**Reply by RobinL** [reply]
Have you been using Splink prior to 4.0.4 without problems?  i.e. do you think this is a bug that has crept in due in this version?  (There was a relevant change in 4.0.4 which was fixed [here](https://github.com/moj-analytical-services/splink/pull/2470) and will be released in 4.0.5

**Reply by Zachary-Higgins-Altera** [reply]
Hi @RobinL - Not sure if you heard back from @pallyndr. We've been evaluating this framework in Databricks (spark) and I'm also having the same issue. Currently on version 4.0.7. I saw in the docs that Databricks support is limited, but I'm very happy to help accommodate some troubleshooting activities if this is something you're interested in resolving.
